### PR TITLE
Add choice param for which buildmaster service or job to deploy

### PR DESCRIPTION
**Pull request by jwbron**
## Summary:
Adds a param to select which job should be deployed. Note, before merging, update the notify channel to 1s-and-0s
Issue: INFRA-10443

## Test plan:
Run as a replay in jenkins

### CODE REVIEW

**Review by nathanjd — APPROVED** [review]

File: jobs/deploy-buildmaster.groovy [hunk 1]
```diff
@@ -21,10 +21,13 @@ new Setup(steps
     "GCLOUD_PROJECT",
     """<b>REQUIRED</b>. The Google Cloud project to deploy to.""",
     ["khan-test", "khan-internal-services"],
-
+).addChoiceParam(
```
Comment (csilvers) [inline jobs/deploy-buildmaster.groovy:24]:
```suggestion

).addChoiceParam(
```
(You need a blank line!)

File: jobs/deploy-buildmaster.groovy [hunk 2]
```diff
@@ -37,10 +40,37 @@ def buildAndDeploy() {
...
-        // Call make deploy with the correct project and branch
-        sh("make deploy");
-        echo("Deployment successful!");
+      withEnv([
+          "GCLOUD_PROJECT=${params.GCLOUD_PROJECT}"
+      ]) {
+        try {
+          // Deploy based on the selected service or job
+          def serviceOrJob = params.SERVICE_OR_JOB
+          switch (serviceOrJob) {
```
Comment (csilvers) [inline jobs/deploy-buildmaster.groovy:49]:
```suggestion
          switch (params.SERVICE_OR_JOB) {
```

File: jobs/deploy-buildmaster.groovy [hunk 3]
```diff
@@ -37,10 +40,37 @@ def buildAndDeploy() {
...
+              sh "make deploy_reaper"
+              break
+            case "trigger-reminders":
+              sh "make deploy_trigger_reminders"
+              break
+            case "warm-rollback":
+              sh "make deploy_warm_rollback"
+              break
+            default:
+              error("Invalid service or job selected: ${serviceOrJob}")
```
Comment (csilvers) [inline jobs/deploy-buildmaster.groovy:66]:
```suggestion
              error("Invalid service or job selected: ${params.SERVICE_OR_JOB}")
```

File: jobs/deploy-buildmaster.groovy [hunk 4]
```diff
@@ -37,10 +40,37 @@ def buildAndDeploy() {
...
-        echo("Deployment successful!");
+      withEnv([
+          "GCLOUD_PROJECT=${params.GCLOUD_PROJECT}"
+      ]) {
+        try {
+          // Deploy based on the selected service or job
+          def serviceOrJob = params.SERVICE_OR_JOB
+          switch (serviceOrJob) {
+            case "all":
+              sh "make deploy"
```
Comment (csilvers) [inline jobs/deploy-buildmaster.groovy:51]:
```suggestion
              sh("make deploy");
```
and likewise for all the other sh's.

File: jobs/deploy-buildmaster.groovy [hunk 5]
```diff
@@ -37,10 +40,37 @@ def buildAndDeploy() {
...
+            case "trigger-reminders":
+              sh "make deploy_trigger_reminders"
+              break
+            case "warm-rollback":
+              sh "make deploy_warm_rollback"
+              break
+            default:
+              error("Invalid service or job selected: ${serviceOrJob}")
+          }
+          echo "Deployment successful!"
```
Comment (csilvers) [inline jobs/deploy-buildmaster.groovy:68]:
```suggestion
          echo("Deployment successful!");
```

File: jobs/deploy-buildmaster.groovy [hunk 6]
```diff
@@ -37,10 +40,37 @@ def buildAndDeploy() {
...
+        } catch (Exception e) {
+          echo "Deployment failed"
+          currentBuild.result = 'FAILURE'
+          throw e
+        }
```
Comment (csilvers) [inline jobs/deploy-buildmaster.groovy:73]: Please get rid of this -- this behavior already happens automatically.

**Review by csilvers — CHANGES_REQUESTED** [review]
Great, straigthforward enough!

You made a bunch of changes that were not strictly part of the PR at hand, such as adding the try/except.  I think they're all unnecessary at best.  Request Changes to revert them, and also to follow our function call style.  This is all stuff I mentioned in the last buildmaster review I did as well, which maybe you never saw?

**Review by csilvers — APPROVED** [review]
awesome potato!

File: jobs/deploy-buildmaster.groovy [hunk 7]
```diff
@@ -38,16 +44,34 @@ def buildAndDeploy() {
...
         echo("Deployment successful!");
       }
     }
   }
 }
 
 onMaster('90m') {
-  notify([slack: [channel: '#hack-buildmaster-2024',
+  notify([slack: [channel: '#1s-and-0s-deploys',
```
[the remark this answers is not in the document]
Comment (csilvers) [reply jobs/deploy-buildmaster.groovy:74]: 1s-and-0s-deploys, in practice, is really only useful for the deploy queue.  Everything else gets lost in the noise.  deploy-fastly.groovy sends to a #fastly channel for this reason.  I think that what you have here is fine, but another channel would probably be better.  Who owns buildmaster?  You could also just send it to #infrastructure, though that's kinda a dead channel too.

[the remark this answers is not in the document]
Comment (csilvers) [reply jobs/deploy-buildmaster.groovy:74]: I think that would be fine!

There's also `#infrastructure-deploys`, that @MiguelCastillo set up, maybe that would be a good choice?  It mostly holds warnings now, but we could move this there.  And maybe fastly deploys one day too.

**Pull request merged**